extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -0,0 +1,430 @@
+            // apiCallTimeout.
+            final long readNanos = instructions.readTimeout().plusMillis(100).toNanos();
+            return consumerFuture.get(readNanos, TimeUnit.NANOSECONDS);
+        }
+
+        private boolean isFragment(long fragmentIndex) {
+            return this.fragmentIndex == fragmentIndex;
+        }
+
+        private int requestLength() {
+            return (int) (to - from + 1);
+        }
+
+        private GetObjectRequest getObjectRequest() {
+            return GetObjectRequest.builder()
+                    .bucket(uri.bucket().orElseThrow())
+                    .key(uri.key().orElseThrow())
+                    .range("bytes=" + from + "-" + to)
+                    .build();
+        }
+
+        private String requestStr() {
+            if (uri != null) {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d %s/%s", System.identityHashCode(S3ChannelContext.this),
+                        fragmentIndex, from, to, requestLength(), uri.bucket().orElseThrow(), uri.key().orElseThrow());
+            } else {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d", System.identityHashCode(S3ChannelContext.this),
+                        fragmentIndex, from, to, requestLength());
+            }
+        }
+
+        // --------------------------------------------------------------------------------------------------
+
+        final class Sub implements Subscriber<ByteBuffer>, BiConsumer<Void, Throwable> {
+            private final CompletableFuture<ByteBuffer> localProducer;
+            private ByteBuffer buffer;
+            private Subscription subscription;
+
+            Sub() {
+                localProducer = producerFuture;
+                buffer = bufferPool.take(requestLength());
+                // 1. localProducer succeeds: whenComplete will executed when released
+                // 2. localProducer fails: whenComplete will executed asap
+                // 3. localProducer limbo (not expected): whenComplete will executed when released
+                CompletableFuture.allOf(localProducer, released).whenComplete(this);
+            }
+
+            // -----------------------------------------------------------------------------
+
+            @Override
+            public synchronized void accept(Void unused, Throwable throwable) {
+                bufferPool.give(buffer);
+                buffer = null;
+            }
+
+            // -----------------------------------------------------------------------------
+
+            @Override
+            public synchronized void onSubscribe(Subscription s) {
+                if (subscription != null) {
+                    s.cancel();
+                    return;
+                }
+                subscription = s;
+                subscription.request(Long.MAX_VALUE);
+            }
+
+            @Override
+            public synchronized void onNext(ByteBuffer byteBuffer) {
+                buffer.put(byteBuffer);
```

Comment:
This could NPE in the "limbo" case. Do we care?

Reply:
The likely alternative is to "gracefully" handle when buffer is null, and do `localProducer.completeExceptionally`. Given I don't expect this to happen, I think hard NPE here is more appropriate (in which case, I expect `onError` to be called regardless, and that will do the `localProducer.completeExceptionally`).

Reply:
I think it depends on whether we're expecting log spam that will scare/annoy users.

Reply:
Shouldn't be any log spam afaik.